### PR TITLE
SPD Fixed Determinant (Isochoric Matrices)

### PR DESCRIPTION
**Pull request by kellertuer**
This came as a small idea this morning and might be useful for someone, but it is indeed just a split of the SPDs into “constant determinant shells”.

Still a nice short thing to do.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
Nice! One small observation, in many many places you have "fixed rank" instead of "fixed determinant".

File: docs/src/manifolds/symmetricpositivedefinitefixedrank.md
Comment (mateuszbaran) [inline docs/src/manifolds/symmetricpositivedefinitefixedrank.md:1]:
```suggestion
# [Symmetric positive definite matrices of fixed determinant](@id SymmetricPositiveDefiniteFixedDeterminantSection)
```

**Comment by kellertuer** [comment]
> Nice! One small observation, in many many places you have "fixed rank" instead of "fixed determinant".

oops, yes I always saw the file on the side and was maybe confused by that every now and then

File: docs/src/manifolds/symmetricpositivedefinitefixeddeterminant.md
Comment (mateuszbaran) [inline docs/src/manifolds/symmetricpositivedefinitefixeddeterminant.md:1]:
```suggestion
# [Symmetric positive definite matrices of fixed determinant](@id SymmetricPositiveDefiniteFixedDeterminantSection)
```

**Comment by mateuszbaran** [comment]
Sorry, I didn't mean to write "many" twice :sweat_smile: .

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:33]:
```suggestion
struct SymmetricPositiveDefiniteFixedDeterminant{N,TD<:Real} <: AbstractDecoratorManifold{ℝ}
    d::TD
end
```
I think determinant should be just a field.

**Comment by kellertuer** [comment]
> Sorry, I didn't mean to write "many" twice 😅 .

That‘s completely fine, I also did not mean to write rank 😅

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl]: Yes, you might be right, there is not much reason to ever dispatch on that.

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:61]:
```suggestion
    if det(p) ≉ M.d
```
I think approximate equality should be sufficient.

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:74]:
```suggestion
[`SymmetricPositiveDefiniteFixedDeterminant`](@ref) `M`, i.e. `X` has to be a symmetric matrix of size `(n,n)`
```

Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:74]: The rest of the docstring doesn't seem to correspond to what is actually tested here.

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:80]:
```suggestion
function check_vector(M::SymmetricPositiveDefiniteFixedDeterminant, p, X; atol::Real=sqrt(max_eps(X)), kwargs...)
    if !isapprox(tr(X), 0.0; atol=atol, kwargs...)
```
So that by default it's not an exact comparison with `0`.

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:115]:
```suggestion
    q = project(M::SymmetricPositiveDefiniteFixedDeterminant{n}, p)
    project!(M::SymmetricPositiveDefiniteFixedDeterminant{n}, q, p)
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:118]:
```suggestion
(sub-)manifold of s.p.d. matrices of determinant `M.d` (in place of `q`).
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:135]:
```suggestion
    Y = project(M::SymmetricPositiveDefiniteFixedDeterminant{n}, p, X)
    project!(M::SymmetricPositiveDefiniteFixedDeterminant{n}, Y, p, X)
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:138]:
```suggestion
(sub-)manifold of s.p.d. matrices of determinant `M.d` (in place of `Y`),
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:146]:
```suggestion
    fill!(Y[diagind(Y)], 0)
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:23]:
```suggestion
since for a constant determinant we require that `0 = D\det(p)[Z] = \det(p)\operatorname{tr}(p^{-1}Z)` for all tangent vectors ``Z``.
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:50]:
```suggestion
whether `p` is a [`SymmetricPositiveDefinite`](@ref) matrix (checked by traits) of size `(n, n)`
```

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:74]: Oh because I forgot to update that. Sorry, I thought it was a quick fix but it seems I missed several things.

File: src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:80]: hm, until now we never did that.

Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:80]: We have that in a few places, for example `KendallsShapeSpace`. I think we should move in this direction for other manifolds as well.

Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:80]: hm until now at least my philosophy was: The default from is approx should be fine enough

Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:80]: The default is enough when we don't compare to 0. When we compare to 0, we should IMO have a non-zero atol by default.

Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:81]: Hm, then is approx has a strange default for that case. I would prefer to maybe have one PR that changes the default behaviour of Manifolds.jl when we compare approximately vs zero overall? Otherwise this would just get a bit inconsistent throughout the package for now.

Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefiniteFixedDeterminant.jl:81]: It's already inconsistent but if you prefer we can leave fixing that for later.

**Comment by kellertuer** [comment]
I think its maybe best to stop for today, now project is broken and I do not see why, maybe your fill does not what you thought it should?

**Comment by kellertuer** [comment]
Just learned these matrices are called Isochronic; maybe that is a nicer name than, but have to check that again.

**Comment by mateuszbaran** [comment]
Ah, it doesn't because I forgot `view`. It should be `fill!(view(Y, diagind(Y)), 0)`

**Comment by mateuszbaran** [comment]
We can of course finish this tomorrow.

**Comment by kellertuer** [comment]
I switched that again; Well I want to think about the name, this is quite long, but isochoric would just be the volume preserving case `d=1`. So some thoughts about the name might be nice; and maybe IsochoricMatrices is still fine since the default is d=1.

**Comment by mateuszbaran** [comment]
Do you have a source where such matrices are called isochoric?

**Comment by kellertuer** [comment]
Until now I only found the systems, we had that in a note we worked with this morning, but I have to check for a source whether the matrices are then also called that.

And then we have to decide whether fixed determinant is fine as well.

**Comment by kellertuer** [comment]
Yes it only refers to volume preservation, so determinant 1 https://en.m.wiktionary.org/wiki/isochoric#English; so then the interpretation question is whether it's fine to have a unified volume change as well (det != 1 but fixed).

**Comment by mateuszbaran** [comment]
OK then let's call it IsochoricMatrices.

**Comment by kellertuer** [comment]
I hope in renaming I no-where stumbled into isotropic or isochronic. ;)

**Comment by sethaxen** [comment]
An alternative to avoid choosing a non-standard name that maybe 1 or 2 people will understand would be to add the alias `SPD` for `SymmetricPositiveDefinite` and then define `SPDFixedDeterminant`, which is not so long, and now the name is understandable by everyone.

**Comment by mateuszbaran** [comment]
Hm, actually `SPDFixedDeterminant` seems more understandable to me than `IsochoricMatrices`.

**Comment by kellertuer** [comment]
That's a bit sad, I was quite happy finding something short that is not an abbreviation – since for now we avoided to abbreviate manifolds (sure we have `SVDPoint` and `SPDPoint`), and I feel the property “volume preserving” (a bithidden in greek) is nicer that describing the equation that is added.

But sure, if I am in the minority I can change the name (back).

File: src/manifolds/SPDFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SPDFixedDeterminant.jl:16]:
```suggestion
These matrices are sometimes also called [isochoric](https://en.wiktionary.org/wiki/isochoric), which refers to the interpretation of
```

File: src/manifolds/SPDFixedDeterminant.jl
Comment (mateuszbaran) [inline src/manifolds/SPDFixedDeterminant.jl:17]:
```suggestion
the matrix representing an ellipsoid. All ellipsoids that represent points on this manifold have the same volume.
```

**Review by mateuszbaran — APPROVED** [review]

**Comment by kellertuer** [comment]
I think this is not urgent so it can also wait for the others before we merge (or we merge it but wait before registering a new version).

**Pull request merged**
